### ISSUE TITLE
The results of `lavTestLRT()` with `method = "satorra.2000"` in 0.6-13.1806 can be different from those in 0.6-12 for mutlisample models

### ISSUE THREAD

**Issue by sfcheung** [yrosseel/lavaan#255]
I noticed that the results of `lavTestLRT()` with `method = "satorra.2000"` in 0.6-12 and 0.6-13.1806 are different:

This is from 0.6-13.1806:

```r
# Adapted from https://lavaan.ugent.be/tutorial/groups.html

library(lavaan)
#> This is lavaan 0.6-13.1806
#> lavaan is FREE software! Please report any bugs.

# No between group constrains
HS.model1 <- ' visual  =~ x1 + x2 + x3
              textual =~ x4 + x5 + x6
              speed   =~ x7 + x8 + c(a1, a2) * x9 '
fit1 <- cfa(HS.model1,
           data = HolzingerSwineford1939,
           group = "school",
           test = "satorra.bentler")

# Loadings of x9 constrained to be equal across group
HS.model2 <- ' visual  =~ x1 + x2 + x3
              textual =~ x4 + x5 + x6
              speed   =~ x7 + x8 + c(a1, a1) * x9 '
fit2 <- cfa(HS.model2,
           data = HolzingerSwineford1939,
           group = "school",
           test = "satorra.bentler")

# Loadings of x9:
parameterEstimates(fit1)[c(9, 45), -4]
#>      lhs op rhs group label   est    se     z pvalue ci.lower ci.upper
#> 9  speed =~  x9     1    a1 0.922 0.225 4.104      0    0.482    1.362
#> 45 speed =~  x9     2    a2 1.058 0.165 6.429      0    0.735    1.380
parameterEstimates(fit2)[c(9, 45), -4]
#>      lhs op rhs group label   est    se     z pvalue ci.lower ci.upper
#> 9  speed =~  x9     1    a1 1.015 0.133 7.626      0    0.754    1.276
#> 45 speed =~  x9     2    a1 1.015 0.133 7.626      0    0.754    1.276

out <- lavTestLRT(fit2, fit1, method = "satorra.2000")
out
#> 
#> Scaled Chi-Squared Difference Test (method = "satorra.2000")
#> 
#> lavaan NOTE:
#>     The "Chisq" column contains standard test statistics, not the
#>     robust test that should be reported per model. A robust difference
#>     test is a function of two standard (not robust) statistics.
#>  
#>      Df    AIC    BIC  Chisq Chisq diff Df diff Pr(>Chisq)
#> fit1 48 7484.4 7706.8 115.85                              
#> fit2 49 7482.6 7701.3 116.02    0.21322       1     0.6443
```

This is from 0.6-12:

```r
# Adapted from https://lavaan.ugent.be/tutorial/groups.html

library(lavaan)
#> This is lavaan 0.6-12
#> lavaan is FREE software! Please report any bugs.

# No between group constrains
HS.model1 <- ' visual  =~ x1 + x2 + x3
              textual =~ x4 + x5 + x6
              speed   =~ x7 + x8 + c(a1, a2) * x9 '
fit1 <- cfa(HS.model1,
           data = HolzingerSwineford1939,
           group = "school",
           test = "satorra.bentler")

# Loadings of x9 constrained to be equal across group
HS.model2 <- ' visual  =~ x1 + x2 + x3
              textual =~ x4 + x5 + x6
              speed   =~ x7 + x8 + c(a1, a1) * x9 '
fit2 <- cfa(HS.model2,
           data = HolzingerSwineford1939,
           group = "school",
           test = "satorra.bentler")

# Loadings of x9:
parameterEstimates(fit1)[c(9, 45), -4]
#>      lhs op rhs group label   est    se     z pvalue ci.lower ci.upper
#> 9  speed =~  x9     1    a1 0.922 0.225 4.104      0    0.482    1.362
#> 45 speed =~  x9     2    a2 1.058 0.165 6.429      0    0.735    1.380
parameterEstimates(fit2)[c(9, 45), -4]
#>      lhs op rhs group label   est    se     z pvalue ci.lower ci.upper
#> 9  speed =~  x9     1    a1 1.015 0.133 7.626      0    0.754    1.276
#> 45 speed =~  x9     2    a1 1.015 0.133 7.626      0    0.754    1.276

out <- lavTestLRT(fit2, fit1, method = "satorra.2000")
out
#> Scaled Chi-Squared Difference Test (method = "satorra.2000")
#> 
#> lavaan NOTE:
#>     The "Chisq" column contains standard test statistics, not the
#>     robust test that should be reported per model. A robust difference
#>     test is a function of two standard (not robust) statistics.
#>  
#>      Df    AIC    BIC  Chisq Chisq diff Df diff Pr(>Chisq)
#> fit1 48 7484.4 7706.8 115.85                              
#> fit2 49 7482.6 7701.3 116.02      0.272       1      0.602
```

The adjusted chi-squares are different, while other results are the same. The difference is small. However, I think they are not supposed to be different.

Is it a bug in 0.6-12 that is corrected in 0.6-13? Or is it a bug in 0.6-13?

**Comment by sfcheung** [comment]
I think I figured out the reason.

In 0.9-12, when `method = "satorra.2000`, `scaled.shifted` affects the results, as documented:

```r
# Adapted from https://lavaan.ugent.be/tutorial/groups.html

library(lavaan)
#> This is lavaan 0.6-12
#> lavaan is FREE software! Please report any bugs.

# No between group constrains
HS.model1 <- ' visual  =~ x1 + x2 + x3
              textual =~ x4 + x5 + x6
              speed   =~ x7 + x8 + c(a1, a2) * x9 '
fit1 <- cfa(HS.model1,
           data = HolzingerSwineford1939,
           group = "school",
           test = "satorra.bentler")

# Loadings of x9 constrained to be equal across group
HS.model2 <- ' visual  =~ x1 + x2 + x3
              textual =~ x4 + x5 + x6
              speed   =~ x7 + x8 + c(a1, a1) * x9 '
fit2 <- cfa(HS.model2,
           data = HolzingerSwineford1939,
           group = "school",
           test = "satorra.bentler")

# Loadings of x9:
parameterEstimates(fit1)[c(9, 45), -4]
#>      lhs op rhs group label   est    se     z pvalue ci.lower ci.upper
#> 9  speed =~  x9     1    a1 0.922 0.225 4.104      0    0.482    1.362
#> 45 speed =~  x9     2    a2 1.058 0.165 6.429      0    0.735    1.380
parameterEstimates(fit2)[c(9, 45), -4]
#>      lhs op rhs group label   est    se     z pvalue ci.lower ci.upper
#> 9  speed =~  x9     1    a1 1.015 0.133 7.626      0    0.754    1.276
#> 45 speed =~  x9     2    a1 1.015 0.133 7.626      0    0.754    1.276

lavTestLRT(fit2, fit1, method = "satorra.2000")
#> Scaled Chi-Squared Difference Test (method = "satorra.2000")
#> 
#> lavaan NOTE:
#>     The "Chisq" column contains standard test statistics, not the
#>     robust test that should be reported per model. A robust difference
#>     test is a function of two standard (not robust) statistics.
#>  
#>      Df    AIC    BIC  Chisq Chisq diff Df diff Pr(>Chisq)
#> fit1 48 7484.4 7706.8 115.85                              
#> fit2 49 7482.6 7701.3 116.02      0.272       1      0.602
lavTestLRT(fit2, fit1, method = "satorra.2000", scaled.shifted = TRUE)
#> Scaled Chi-Squared Difference Test (method = "satorra.2000")
#> 
#> lavaan NOTE:
#>     The "Chisq" column contains standard test statistics, not the
#>     robust test that should be reported per model. A robust difference
#>     test is a function of two standard (not robust) statistics.
#>  
#>      Df    AIC    BIC  Chisq Chisq diff Df diff Pr(>Chisq)
#> fit1 48 7484.4 7706.8 115.85                              
#> fit2 49 7482.6 7701.3 116.02      0.272       1      0.602
lavTestLRT(fit2, fit1, method = "satorra.2000", scaled.shifted = FALSE)
#> Scaled Chi-Squared Difference Test (method = "satorra.2000")
#> 
#> lavaan NOTE:
#>     The "Chisq" column contains standard test statistics, not the
#>     robust test that should be reported per model. A robust difference
#>     test is a function of two standard (not robust) statistics.
#>  
#>      Df    AIC    BIC  Chisq Chisq diff Df diff Pr(>Chisq)
#> fit1 48 7484.4 7706.8 115.85                              
#> fit2 49 7482.6 7701.3 116.02    0.21322       1     0.6443
```

The default results use `scaled.shifted = TRUE`, the default, as documented.

However, it seems that this argument is ignored and `scaled.shifted = FALSE` is always used:

```r
# Adapted from https://lavaan.ugent.be/tutorial/groups.html

library(lavaan)
#> This is lavaan 0.6-13.1806
#> lavaan is FREE software! Please report any bugs.

# No between group constrains
HS.model1 <- ' visual  =~ x1 + x2 + x3
              textual =~ x4 + x5 + x6
              speed   =~ x7 + x8 + c(a1, a2) * x9 '
fit1 <- cfa(HS.model1,
           data = HolzingerSwineford1939,
           group = "school",
           test = "satorra.bentler")

# Loadings of x9 constrained to be equal across group
HS.model2 <- ' visual  =~ x1 + x2 + x3
              textual =~ x4 + x5 + x6
              speed   =~ x7 + x8 + c(a1, a1) * x9 '
fit2 <- cfa(HS.model2,
           data = HolzingerSwineford1939,
           group = "school",
           test = "satorra.bentler")

# Loadings of x9:
parameterEstimates(fit1)[c(9, 45), -4]
#>      lhs op rhs group label   est    se     z pvalue ci.lower ci.upper
#> 9  speed =~  x9     1    a1 0.922 0.225 4.104      0    0.482    1.362
#> 45 speed =~  x9     2    a2 1.058 0.165 6.429      0    0.735    1.380
parameterEstimates(fit2)[c(9, 45), -4]
#>      lhs op rhs group label   est    se     z pvalue ci.lower ci.upper
#> 9  speed =~  x9     1    a1 1.015 0.133 7.626      0    0.754    1.276
#> 45 speed =~  x9     2    a1 1.015 0.133 7.626      0    0.754    1.276

lavTestLRT(fit2, fit1, method = "satorra.2000")
#> 
#> Scaled Chi-Squared Difference Test (method = "satorra.2000")
#> 
#> lavaan NOTE:
#>     The "Chisq" column contains standard test statistics, not the
#>     robust test that should be reported per model. A robust difference
#>     test is a function of two standard (not robust) statistics.
#>  
#>      Df    AIC    BIC  Chisq Chisq diff Df diff Pr(>Chisq)
#> fit1 48 7484.4 7706.8 115.85                              
#> fit2 49 7482.6 7701.3 116.02    0.21322       1     0.6443
lavTestLRT(fit2, fit1, method = "satorra.2000", scaled.shifted = TRUE)
#> 
#> Scaled Chi-Squared Difference Test (method = "satorra.2000")
#> 
#> lavaan NOTE:
#>     The "Chisq" column contains standard test statistics, not the
#>     robust test that should be reported per model. A robust difference
#>     test is a function of two standard (not robust) statistics.
#>  
#>      Df    AIC    BIC  Chisq Chisq diff Df diff Pr(>Chisq)
#> fit1 48 7484.4 7706.8 115.85                              
#> fit2 49 7482.6 7701.3 116.02    0.21322       1     0.6443
lavTestLRT(fit2, fit1, method = "satorra.2000", scaled.shifted = FALSE)
#> 
#> Scaled Chi-Squared Difference Test (method = "satorra.2000")
#> 
#> lavaan NOTE:
#>     The "Chisq" column contains standard test statistics, not the
#>     robust test that should be reported per model. A robust difference
#>     test is a function of two standard (not robust) statistics.
#>  
#>      Df    AIC    BIC  Chisq Chisq diff Df diff Pr(>Chisq)
#> fit1 48 7484.4 7706.8 115.85                              
#> fit2 49 7482.6 7701.3 116.02    0.21322       1     0.6443
```

Is it an intended change to be documented? Or is it a bug in 0.9-13.1806?

**Comment by sfcheung** [comment]
P.S.: I did not install 0.9-13.1806 intentionally. I installed a GitHub version of lavaan in a copy of R a few days ago. When I installed something else, I was prompted to update lavaan (maybe by a function of devtools?) and then I ended up with this version of lavaan, according to the message when I load lavaan.

**Comment by yrosseel** [comment]
This is the result of a bug in 0.6-12 (and all previous versions), that was finally found and fixed in 0.6-13. So the 0.6-13 result is correct.

There is only a bug (in <0.6-13) if 1) a mean and variance adjusted test is used (i.e., test = “scaled.shifted”, or test = “mean.var.adjusted”), 2) multiple group sare involved, and 3) equality constraints (across groups) are involved.

The scaled.shifted= argument still works, but just happens to produce the same result in this example. If you add, for example, group.equal = "loadings" to the second model, you will see a (small) difference depending on whether scaled.shiftted is TRUE or FALSE.

**Comment by sfcheung** [comment]
Maybe there were actually two bugs in 0.6-12?

As shown in an example I posted at the Google Group, the difference between 0.6-12 and 0.6-13 can be nonnegligible, even without between-group equality constraints:

https://groups.google.com/g/lavaan/c/zZokm8M7fC4/m/kjrZ0HK7BAAJ

I tried to understand the change, and comments on the change led me to this in the old approach (pre 0.6-12):

https://github.com/yrosseel/lavaan/blob/43b8be94a7adeff6c2c35f8693f0d7fc07a1273d/R/lav_test_diff.R#L140-L142

Should `fg` be replaced by `fg^2`?:

```r
trace.UGamma2 <- sum(fg^2 * trace.UGamma2)
```

I created a test branch, added arguments to allow (a) selecting the approach to use (0.6-13 or pre-0.6-13) and (b) whether `fg` or `fg^2` is used in the pre-0.6-13 approach:

https://github.com/sfcheung/lavaan/tree/satorra2000_fix_old

When calling `lavTestLRT()`, if `old.approach = TRUE`, the pre-0.6-13 approach will be used. If `fg2` is also `TRUE`, then `fg^2` will be used.

The following example is based on this branch:

```r
# Adapted from https://lavaan.ugent.be/tutorial/groups.html

#> ℹ Loading lavaan
#> This is lavaan 0.6-14 <-- Ignore this
#> lavaan is FREE software! Please report any bugs.
set.seed(65145)
dat <- HolzingerSwineford1939
dat <- dat[sample(nrow(dat), 200), ]

# No between group constrains
HS.model1 <- ' visual  =~ x1 + x2 + x3
              textual =~ x4 + x5 + x6
              speed   =~ x7 + x8 + x9
              visual ~~ textual
              visual ~~ speed
              textual ~~ speed'
fit1 <- cfa(HS.model1,
           data = dat,
           group = "school",
           meanstructure = TRUE,
           test = "satorra.bentler",
           check.start = FALSE)

HS.model2 <- ' visual  =~ x1 + x2 + x3
              textual =~ x4 + x5 + x6
              speed   =~ x7 + x8 + x9
              visual ~~ textual
              visual ~~ c(0, NA) * speed
              textual ~~ speed'
fit2 <- cfa(HS.model2,
           data = dat,
           group = "school",
           meanstructure = TRUE,
           test = "satorra.bentler",
           start = fit1,
           check.start = FALSE)

parameterestimates(fit1)[11, ]
#>       lhs op   rhs block group   est    se     z pvalue ci.lower ci.upper
#> 11 visual ~~ speed     1     1 0.191 0.098 1.955  0.051   -0.001    0.384
parameterestimates(fit2)[11, ]
#>       lhs op   rhs block group est se  z pvalue ci.lower ci.upper
#> 11 visual ~~ speed     1     1   0  0 NA     NA        0        0

lavTestLRT(fit2, fit1, method = "satorra.2000")
#> 
#> New approach used.
#> 
#> Scaled Chi-Squared Difference Test (method = "satorra.2000")
#> 
#> lavaan NOTE:
#>     The "Chisq" column contains standard test statistics, not the
#>     robust test that should be reported per model. A robust difference
#>     test is a function of two standard (not robust) statistics.
#>  
#>      Df    AIC    BIC  Chisq Chisq diff Df diff Pr(>Chisq)  
#> fit1 48 5008.1 5206.0 109.64                                
#> fit2 49 5010.6 5205.2 114.14     4.8927       1    0.02697 *
#> ---
#> Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

lavTestLRT(fit2, fit1, method = "satorra.2000", old.approach = FALSE)
#> 
#> New approach used.
#> 
#> Scaled Chi-Squared Difference Test (method = "satorra.2000")
#> 
#> lavaan NOTE:
#>     The "Chisq" column contains standard test statistics, not the
#>     robust test that should be reported per model. A robust difference
#>     test is a function of two standard (not robust) statistics.
#>  
#>      Df    AIC    BIC  Chisq Chisq diff Df diff Pr(>Chisq)  
#> fit1 48 5008.1 5206.0 109.64                                
#> fit2 49 5010.6 5205.2 114.14     4.8927       1    0.02697 *
#> ---
#> Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
lavTestLRT(fit2, fit1, method = "satorra.2000", old.approach = TRUE, fg2 = TRUE)
#> 
#> Old approach used.
#> Use 'trace.UGamma2 <- sum(fg^2 * trace.UGamma2)'
#> 
#> Scaled Chi-Squared Difference Test (method = "satorra.2000")
#> 
#> lavaan NOTE:
#>     The "Chisq" column contains standard test statistics, not the
#>     robust test that should be reported per model. A robust difference
#>     test is a function of two standard (not robust) statistics.
#>  
#>      Df    AIC    BIC  Chisq Chisq diff Df diff Pr(>Chisq)  
#> fit1 48 5008.1 5206.0 109.64                                
#> fit2 49 5010.6 5205.2 114.14     4.8927       1    0.02697 *
#> ---
#> Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
lavTestLRT(fit2, fit1, method = "satorra.2000", old.approach = TRUE, fg2 = FALSE)
#> 
#> Old approach used.
#> Use 'trace.UGamma2 <- sum(fg * trace.UGamma2)'
#> 
#> Scaled Chi-Squared Difference Test (method = "satorra.2000")
#> 
#> lavaan NOTE:
#>     The "Chisq" column contains standard test statistics, not the
#>     robust test that should be reported per model. A robust difference
#>     test is a function of two standard (not robust) statistics.
#>  
#>      Df    AIC    BIC  Chisq Chisq diff Df diff Pr(>Chisq)  
#> fit1 48 5008.1 5206.0 109.64                                
#> fit2 49 5010.6 5205.2 114.14     3.8339       1    0.05023 .
#> ---
#> Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
```

Using `fg^2`, the old approach and the correct 0.6-13 approach yield the same results, as expected.

Not presented above but I also tried to impose equality constraints and see what happen. If `fg^2` is used in the old approach, the old and new approach yield different results but, as expected, the difference is small.

Even if this is indeed a bug (another bug) in 0.6-12 or older version, it has already been fixed in 0.6-13, and so there is no need to fix it. Nevertheless, if this is indeed a bug, then this can explain the nonnegligible difference found in some cases between 0.6-13 and previous version in this test.

**Comment by falkcarl** [comment]
Not sure if it's helpful, but there was another time in the past when difference tests had broken across lavaan versions (I also wanted user specified starting values to not interfere with how the tests were computed or work in a particular way, something like that) and I wanted testing code to ensure I would at least know if something changed. That was apparently May 9, 2018, and on a lavaan fork: 0401880f5a049c84fc6a5220b5ee415634934af2

The current development version at least passes the tests. How I verified that the tests were indeed correct could take more work, however, as my sandbox code that I used to manually compute the difference tests didn't work its way into a git repo. I do not recall testing or working with these in the context of multiple group models.

**Comment by sfcheung** [comment]
@falkcarl , thanks for sharing the tests. having tests is a good idea, for identifying expected and unexpected changes. However, I understand it may involve a lot of work to add them to `lavaan` (or maybe there are already such tests?). Moreover, CRAN has a limit on the time for running the tests.

Maybe a fork of `lavaan` or a separate "package" that is devoted to testing results from `lavaan`?

P.S.: I noticed that the title I used for this issue can be misleading. The discrepancies occurred only for multisample models. As your tests demonstrated, the results for single-sample models are the same across recent versions (0.6-12 and 0.6-13). I amended the title of this issue.
